server/topic.go

```diff
@@ -577,10 +591,33 @@ func (t *Topic) run(hub *Hub) {
 				}
 			}
 		case ua := <-t.uaChange:
-			// process an update to user agent from one of the sessions
+			// Process an update to user agent from one of the sessions
 			currentUA = ua
 			uaTimer.Reset(uaTimerDelay)
 
+		case <-defrNotifTimer.C:
+			// Handle deferred presence notifications from a successful service (background) subscription.
+			if !t.isReady() {
+				continue
+			}
+
+			// Process events older than this timestamp.
+			expiration := time.Now().Add(-deferredNotificationsTimeout)
+			// Iterate through the list until all sufficiently old events are processed.
+			for elem := t.defrNotif.Back(); elem != nil; elem = t.defrNotif.Back() {
+				sreg := elem.Value.(*sessionJoin)
+				if expiration.Before(sreg.pkt.timestamp) {
+					// All done. Remaining events are newer.
+					break
+				}
+				t.defrNotif.Remove(elem)
+				if pssd, ok := t.sessions[sreg.sess]; ok {
+					pssd.ref = nil
+					t.sessions[sreg.sess] = pssd
+				}
+				t.sendSubNotifications(types.ParseUserId(sreg.pkt.from), sreg)
```

Comment:
Should we also update online counters?

Reply:
Online counters are updated here:
https://github.com/tinode/chat/blob/sync/server/topic.go#L1198
https://github.com/tinode/chat/blob/sync/server/topic.go#L239

And I added one more case where it needs to be updated.

Reply:
https://github.com/tinode/chat/blob/sync/server/topic.go#L1201 increments the counter for foreground sessions only when a session subscribes to the topic.

This piece of code sends presences notifications if the session was subscribed in the background but didn't go away within `deferredNotificationsTimeout` seconds.
So sending `{pres}` means we are promoting the session to foreground but the counter won't be incremented. Am I missing something?

Reply:
You are correct and I fixed it here: https://github.com/tinode/chat/pull/323/commits/4eaaa40c639471a8d937e0d0a8538bb0f19fef33